<a href="https://colab.research.google.com/github/manassanant132/Fundamental-of-Deep-Learning/blob/main/Unit11_%E0%B8%81%E0%B8%B2%E0%B8%A3%E0%B9%80%E0%B8%A5%E0%B8%B7%E0%B8%AD%E0%B8%81%E0%B9%83%E0%B8%8A%E0%B9%89_Loss_Function_%E0%B9%83%E0%B8%99%E0%B8%81%E0%B8%B2%E0%B8%A3%E0%B8%9E%E0%B8%B1%E0%B8%92%E0%B8%99%E0%B8%B2_Deep_Learning_Model_(%E0%B8%95%E0%B8%AD%E0%B8%99%E0%B8%97%E0%B8%B5%E0%B9%88_2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ตอนที่ 2 ของบทความเรื่องการเลือกใช้ Loss Function ในการพัฒนา Deep Learning Model นี้ ผู้อ่านจะได้ทำ Workshop ที่มีการคอนฟิก Model แบบ Classification ด้วย Loss Function อีก 3 ตัว ได้แก่



*   Binary Crossentropy Loss

*   Categorical Crossentropy Loss

*   Sparse Categorical Crossentropy Loss

แต่ก่อนอื่นเราจะทำความเข้าใจแนวคิดของ Information, Entropy และ Cross-Entropy  ซึ่งเป็นพื้นฐานสำคัญของ Loss Function ทั้ง 3 ตัวกันก่อนครับ





#Information Theory

Information Theory ถูกคิดค้นขึ้นโดย Claude Shannon วิศวกรไฟฟ้าและนักคณิตศาสตร์ชาวอเมริกัน ในปี พ.ศ.2491 ขณะที่เขาทำงานอยู่ที่ Bell Labs บริษัทโทรคมนาคมขนาดใหญ่ โดย Information Theory เป็นศาสตร์ที่ศึกษาเกี่ยวกับการหาปริมาณ Information การบีบอัดข้อมูล และขีดจำกัดของการประมวลผลสัญญาณ ฯลฯ

ในการหาปริมาณของ Information จะใช้ความเข้าใจในเรื่องของเหตุการณ์ (Event) ตัวแปรสุ่ม (Random Variable) และการแจกแจงความน่าจะเป็น (Probability Distribution) ฯลฯ ซึ่งจะทำให้เราสามารถวัดปริมาณ Information จาก Data ที่มีการสื่อสารกัน โดยใช้แนวคิดที่ว่า

"เหตุการณ์ที่มีโอกาสเกิดขึ้นต่ำ (Low Probability) จะมีความน่า Surprise หรือมี Information (มูลค่า) สูง ในขณะที่เหตุการณ์ที่มีโอกาสเกิดขึ้นสูง (High Probability) จะมี Information (มูลค่า) ต่ำ"


ตามสูตรดังต่อไปนี้

information(x) = -log(p(x))

เราสามารถแทน information(x) ด้วย h(x) ได้เช่นกัน

In [ ]:
import math
def p(x):
  return x

-math.log2(p(0.7))

0.5145731728297583

โดยที่ log() คือ Logarithm ฐาน 2 และ p(x) คือความน่าจะเป็นที่จะเกิดเหตุการณ์ x

ยกตัวอย่างในภาษาไทยของเราที่จะมีโอกาสพบคำว่า อาตมา ตามที่ต่างๆ น้อยกว่าคำว่า ผม และ ฉัน ซึ่งใช้แทนตัวผู้พูดโดยไม่เจาะจงว่าเป็นใคร แต่เมื่อมีการพบคำว่า อาตมา ในประโยค เราก็จะเข้าใจได้ในทันทีว่าผู้พูดเป็นบุคคลพิเศษ ไม่ได้ประกอบอาชีพทั่วไป ดังนั้นเหตุการณ์ที่เกิดคำว่า อาตมา จึงมี Information สูงว่าเหตุการณ์ที่จะพบคำว่า ผม และ ฉัน

เนื่องจากเราใช้ Logarithm ฐาน 2 หน่วยของ Information จึงเป็นจำนวน bit (Binary Digit) ที่จำเป็นในการแสดงถึงเหตุการณ์หนึ่งๆ

In [ ]:
from math import log, log2
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

import tensorflow as tf

In [ ]:
p = 0.1
h = -log2(p)

print('p(x)=%.3f, information: %.3f bits' % (p, h))

p = 0.5
h = -log2(p)

print('p(x)=%.3f, information: %.3f bits' % (p, h))

p(x)=0.100, information: 3.322 bits
p(x)=0.500, information: 1.000 bits


p(x)=0.100, information: 3.322 bits
p(x)=0.500, information: 1.000 bits

ดังนั้น ยิ่งเหตุการณ์มีโอกาสเกิดขึ้นได้สูง ความ Surprise หรือ Information หรือจำนวน bit ที่จำเป็นในการแสดงถึงเหตุการณ์หนึ่งๆ จะยิ่งต่ำ ดังกราฟด้านล่าง

In [ ]:
probs = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]

info = [-log2(p) for p in probs]

In [ ]:
df = pd.DataFrame(list(zip(probs, info)), columns =['Probability', 'Information'])
df

Probability  Information
0          0.1     3.321928
1          0.2     2.321928
2          0.3     1.736966
3          0.4     1.321928
4          0.5     1.000000
5          0.6     0.736966
6          0.7     0.514573
7          0.8     0.321928
8          0.9     0.152003
9          1.0    -0.000000

In [ ]:
fig = px.line(df, x='Probability', y='Information', title='Probability vs Information')
fig.show()

เราสามารถนำ Information Theory ไปประยุกต์ใช้ในการบีบอัดข้อมูลเพื่อประหยัดค่าใช้จ่ายในการสื่อสาร โดยสมมติว่าถ้าเรามีตัวอักษรอยู่ 4 ตัว ได้แก่ 'A', 'B', 'C' และ 'D' ซึ่งแต่ละตัวมีโอกาสปรากฏอยู่ในข้อความ (Text) ด้วยความน่าจะเป็น p เท่ากับ 1/2, 1/4, 1/8, และ 1/8 ตามลำดับ (รวมกันเท่ากับ 1.0) ดังนั้นเพื่อจะประหยัดจำนวน bit ในการส่งข้อมูล เราจะเข้ารหัสตัวอักษรทั้ง 4 ตัว ดังต่อไปนี้

#Entropy

ดังนั้น เราสามารถคำนวณหาจำนวน bit โดยเฉลี่ยที่จำเป็นในการส่งข้อมูล สำหรับตัวอักษร 4 ตัว ที่มีการแจกแจงความน่าจะเป็น (Probability Distribution) เท่ากับ 1/2, 1/4, 1/8, และ 1/8 ได้ดังนี้

Average bit หรือ H(X) = (1/2 × 1) +  (1/4  × 2) + (1/8  × 3) + (1/8  × 3) = 1.75

เราเรียกจำนวน bit โดยเฉลี่ยดังกล่าวว่า Entropy (หรือ H(X) โดย X คือ Random Variable) ซึ่งคล้ายกับแนวคิดของ Entropy ที่เป็นระบบทางกายภาพในวิชาฟิสิกส์ โดยทั้ง 2 จะให้ความสำคัญกับเรื่องของความไม่แน่นอน (Uncertainty)

สมมติว่ามีการโยนลูกเต๋า 1 ลูก ที่มีโอกาสออกแต่ละหน้าด้วยความน่าจะเป็น 1/6 (Uniform Probability Distribution) เราสามารถคาดหวังได้ว่าจำนวน bit โดยเฉลี่ย หรือ Entropy จะมีค่าดังนี้

In [ ]:
P = [1/6, 1/6, 1/6, 1/6, 1/6, 1/6]

entropy = -sum([p * log2(p) for p in P])

print('entropy: %.3f bits' % entropy)

entropy: 2.585 bits


ในอีกแง่หนึ่ง Entropy นั้นบอกถึงความ Surprise โดยเฉลี่ย ซึ่งขึ้นอยู่กับลักษณะของการแจกแจงความน่าจะเป็น (Probability Distribution) ของเหตุการณ์หนึ่งๆ ดังตัวอย่างด้านล่าง

In [ ]:
def entropy(events, ets=1e-15):
    return -sum([p * log2(p + ets) for p in events])

probs = [0.0, 0.1, 0.2, 0.3, 0.4, 0.5]

dists = [[p, 1.0 - p] for p in probs]

ents = [entropy(d) for d in dists]

In [ ]:
df = pd.DataFrame(list(zip(probs, ents)), columns =['Probability Distribution', 'Entropy (bits)'])

In [ ]:
fig = px.line(df, x='Probability Distribution', y='Entropy (bits)', title='Probability Distribution vs Entropy')
fig.update_xaxes(ticktext=[str(d) for d in dists], tickvals = probs)

ซึ่งจากกราฟจะเห็นว่ายิ่งมีการกระจายตัวของความน่าจะเป็นที่สมดุล (Balanced Probability Distribution) ตามแกน x มากขึ้นเท่าไหร่ ความ Surprise หรือ Entropy ก็จะยิ่งมากขึ้นเท่านั้น

ยกตัวอย่างเช่น ในการโยนเหรียญที่มีโอกาสออกหัว 0% และออกก้อย 100% (Skewed Probability Distribution) ซึ่งไม่ว่าจะโยนกี่ครั้งมันก็จะออกก้อยตลอด ดังนั้นความ Surprise หรือ Entropy จึงเท่ากับ 0 ในขณะที่อีกกรณีหนึ่ง ถ้าในการโยนเหรียญมีโอกาสออกหัว 50% และออกก้อย 50% เท่ากัน (Balanced Probability Distribution) ความ Surprise หรือ Entropy จะมีค่าสูงสุดเท่ากับ 1 ครับ

#Cross-Entropy

ทีนี้สมมติว่าในเวลาต่อมา การแจกแจงความน่าจะเป็น (Probability Distribution) ของตัวอักษร 4 ตัว ได้แก่ 'A', 'B', 'C' และ 'D' เปลี่ยนไปจากแบบ p = [1/2, 1/4, 1/8, 1/8] เป็นแบบ q = [1/8, 1/4, 1/2, 1/8]

In [ ]:
events = ['A', 'B', 'C', 'D']
p = [1/2, 1/4, 1/8, 1/8]
q = [1/8, 1/4, 1/2, 1/8]

In [ ]:
fig = go.Figure(data=[
    go.Bar(name='Probability Distribution เก่า (p)', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Probability Distribution ใหม่ (q)', x=events, y=q, text=q, textposition='auto')
])

fig.update_layout(barmode='group')

หากยังมีการเข้ารหัสตัวอักษรทั้ง 4 ตัว แบบเดิม จำนวน bit โดยเฉลี่ย ที่จำเป็นในการส่งข้อมูลจะเพิ่มขึ้นจาก 1.75 bit เป็น 2.5 bit!

Average bit หรือ H(p,q) = (1/8 × 1) +  (1/4  × 2) + (1/2  × 3) + (1/8  × 3) = 2.5

เราเรียกจำนวน bit โดยเฉลี่ย เมื่อมีการเข้ารหัสตัวอักษรแบบเดิม (p) ด้วยการแจกแจงความน่าจะเป็นใหม่ (q) นี้ว่า Cross-Entropy

In [ ]:
def cross_entropy(p, q):
    return -sum([q[i]*log2(p[i]) for i in range(len(p))])

In [ ]:
ce = cross_entropy(p, q)
print('H(P, Q): %.3f bits' % ce)

H(P, Q): 2.500 bits


Cross-Entropy ถูกใช้ในการประมาณ Error ของ Classification Model ที่เกิดจากการแจกแจงความน่าจะเป็น 2 แบบ คือ ระหว่าง p กับ q ซึ่ง q คือ การแจกแจงความน่าจะเป็นที่เราอยากได้ และ p คือ การแจกแจงความน่าจะเป็นที่ถูกประมาณโดย Model

*H(P, Q) != H(Q, P)

#Binary Classification Loss Functions

Binary Classification เป็น Model ที่มีการกำหนด Label หรือ Class เพียง 2 Class คือ ไม่เป็น Class 0 ก็เป็น Class 1 โดย ผลลัพธ์จากการทำนายของ Model จะบอกค่าความน่าจะเป็นว่ามีโอกาสที่จะเป็น Class 1 กี่เปอร์เซ็นต์

เราจะทดลอง Train Binary Classification Model โดยใช้ Binary Crossentropy Loss ของ Keras Framework ด้วย Dataset ที่ Make ขึ้นจากฟังก์ชัน make_circles ของ sklearn Library

*เพื่อจะทำให้ Model ทำนายผลออกมาเป็นค่าความน่าจะเป็น [0, 1] ว่ามีโอกาสที่จะเป็น Class 1 กี่เปอร์เซ็นต์ เราจะต้องคอนฟิก Activate Function ใน Output Layer แบบ Sigmoid

In [ ]:
def sigmoid(x):
    return 1.0/(1+ np.exp(-x))

data = 25
result = sigmoid(data)

print(result)

0.999999999986112


##Binary Crossentropy Loss


In [ ]:
events = ['Class 0', 'Class 1']
actual = [1, 0, 1, 1, 0]
predicted = [0.7, 0.45, 0.9, 0.5, 0.3]

index = 0
p = [1-actual[index], actual[index]]
q = [1-predicted[index], predicted[index]]

In [ ]:
print(p)
print(q)

[0, 1]
[0.30000000000000004, 0.7]


In [ ]:
(1-log2(0.7))+(0-log2(0.3))

3.2515387669959646

In [ ]:
fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=list(np.round(q,2)), textposition='auto')
])

fig.update_layout(barmode='group', title='Record 1')
#0.8812908992306927

In [ ]:
index = 1
p = [1-actual[index], actual[index]]
q = [1-predicted[index], predicted[index]]

In [ ]:
print(p)
print(q)

[1, 0]
[0.55, 0.45]


In [ ]:
(1+log2(0.55))+(0+log2(0.45))

-1.014499569695115

In [ ]:
fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=q, textposition='auto')
])

fig.update_layout(barmode='group', title='Record 2')
#0.9927744539878083

In [ ]:
index = 2
p = [1-actual[index], actual[index]]
q = [1-predicted[index], predicted[index]]

In [ ]:
print(p)
print(q)

[0, 1]
[0.09999999999999998, 0.9]


In [ ]:
#(1+log2(1))+(0+log2(0))

In [ ]:
fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=np.round(q,2), textposition='auto')
])

fig.update_layout(barmode='group', title='Record 3')
#0.4689955935892812

In [ ]:
index = 3
p = [1-actual[index], actual[index]]
q = [1-predicted[index], predicted[index]]

In [ ]:
fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=q, textposition='auto')
])

fig.update_layout(barmode='group', title='Record 4')
#1

In [ ]:
index = 4
p = [1-actual[index], actual[index]]
q = [1-predicted[index], predicted[index]]

In [ ]:
fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=q, textposition='auto')
])

fig.update_layout(barmode='group', title='Record 5')
#0.8812908992306927

Binary Crossentropy คือ ค่าเฉลี่ยของ Cross-Entropy ที่เกิดจากการแจกแจงความน่าจะเป็น 2 แบบ คือ การแจกแจงความน่าจะเป็นที่เราอยากได้ (Actual) กับการแจกแจงความน่าจะเป็นที่ถูกประมาณโดย Model (Predicted) ของ Class 0 และ Class 1 ดังภาพด้านบน ซึ่งการได้ค่าเฉลี่ยน้อยนั้นดีกว่าการได้ค่าเฉลี่ยมาก โดย Binary Crossentropy Loss จะให้ค่าเฉลี่ยต่ำสุดคือ 0

อย่างไรก็ตาม Binary Crossentropy Loss ของ Keras จะมีการใช้งาน Logarithm ฐาน e แทนที่จะเป็น Logarithm ฐาน 2 เหมือนกับตัวอย่างที่ผ่านมาครับ

In [ ]:
def binary_cross_entropy_loss(actual, predicted):
    sum = 0
    for i in range(len(actual)):
        sum=sum+actual[i]*log(predicted[i])+(1-actual[i])*log(1-predicted[i])

    return -sum/len(actual)

binary_cross_entropy_loss(actual, predicted)

0.4219389169701714

In [ ]:
loss = tf.keras.losses.binary_crossentropy(actual, predicted)
loss.numpy()

0.42193872

In [ ]:
actual = [1, 0, 1, 1, 0]
predicted = [1.0, 0.0, 1.0, 1.0, 0.0]

loss = tf.keras.losses.binary_crossentropy(actual, predicted)
loss.numpy()

0.0

###Example

เราจะทดลอง Train Model แบบ Binary Classification จากข้อมูลที่ Make ขึ้นมาด้วยฟังก์ชัน make_circles ตามขั้นตอนต่อไปนี้

Import Library ที่จำเป็นต้องใช้ในการทดลอง

In [ ]:
to_categorical = tf.keras.utils.to_categorical
from tensorflow.keras.utils import to_categorical

import plotly
import plotly.graph_objs as go
import plotly.express as px

from sklearn.datasets import make_circles
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split

import pandas as pd

สร้าง Dataset แบบ 2 Class โดยใช้ Function make_circles ของ Sklearn

In [ ]:
x, y =  make_circles(n_samples=5000, noise=0.1, random_state=1)

แบ่งข้อมูลสำหรับ Train และ Test โดยการสุ่มในสัดส่วน 50:50

In [ ]:
x_train, x_val, y_train, y_val = train_test_split(x, y, test_size=0.4, shuffle= True)

x_train.shape, x_val.shape, y_train.shape, y_val.shape

((3000, 2), (2000, 2), (3000,), (2000,))

นำ Dataset ส่วนที่ Train มาแปลงเป็น DataFrame โดยเปลี่ยนชนิดข้อมูลใน Column "class" เป็น String เพื่อทำให้สามารถแสดงสีแบบไม่ต่อเนื่องได้ แล้วนำไป Plot

In [ ]:
x_train_pd = pd.DataFrame(x_train, columns=['x', 'y'])
y_train_pd = pd.DataFrame(y_train, columns=['class'], dtype='str')

df = pd.concat([x_train_pd, y_train_pd], axis=1)

In [ ]:
fig = px.scatter(df, x="x", y="y", color="class")
fig.show()

นิยาม Model โดยกำหนด Activation Function ใน Layer สุดท้ายเป็น sigmoid

In [ ]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(60, input_dim=2, activation='relu', kernel_initializer='he_uniform'))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(1, activation='sigmoid')) #sigmoid

Compile Model โดยกำหนด Loss Function เป็น binary_crossentropy

In [ ]:
opt = tf.keras.optimizers.SGD(lr=0.01, momentum=0.9)
model.compile(loss='binary_crossentropy', optimizer=opt, metrics=['accuracy'])

Train Model

In [ ]:
his = model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=2000, verbose=1, batch_size = 256)

Epoch 1/2000
12/12 [==============================] - 1s 15ms/step - loss: 0.7355 - accuracy: 0.4970 - val_loss: 0.6975 - val_accuracy: 0.5140
Epoch 2/2000
12/12 [==============================] - 0s 6ms/step - loss: 0.7214 - accuracy: 0.4977 - val_loss: 0.6852 - val_accuracy: 0.5345
Epoch 3/2000
12/12 [==============================] - 0s 4ms/step - loss: 0.7022 - accuracy: 0.5173 - val_loss: 0.6797 - val_accuracy: 0.5395
Epoch 4/2000
12/12 [==============================] - 0s 5ms/step - loss: 0.6982 - accuracy: 0.5177 - val_loss: 0.6751 - val_accuracy: 0.5440
Epoch 5/2000
12/12 [==============================] - 0s 4ms/step - loss: 0.6949 - accuracy: 0.5283 - val_loss: 0.6724 - val_accuracy: 0.5165
Epoch 6/2000
12/12 [==============================] - 0s 5ms/step - loss: 0.6831 - accuracy: 0.5437 - val_loss: 0.6691 - val_accuracy: 0.5450
Epoch 7/2000
12/12 [==============================] - 0s 4ms/step - loss: 0.6820 - accuracy: 0.5523 - val_loss: 0.6655 - val_accuracy: 0.5895
Epoch

Plot Loss

In [ ]:
h1 = go.Scatter(y=his.history['loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="loss"
                   )
h2 = go.Scatter(y=his.history['val_loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_loss"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Loss',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

Plot Accuracy

In [ ]:
h1 = go.Scatter(y=his.history['accuracy'],
                    mode="lines", line=dict(
                    width=2,
                    color='blue'),
                    name="acc"
                   )
h2 = go.Scatter(y=his.history['val_accuracy'],
                    mode="lines", line=dict(
                    width=2,
                    color='red'),
                    name="val_acc"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Accuracy',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

Evaluate Model

In [ ]:
_, train_acc = model.evaluate(x_train, y_train, verbose=0)
_, val_acc = model.evaluate(x_val, y_val, verbose=0)
print('Train: %.4f, Validation: %.4f' % (train_acc, val_acc))

Train: 0.8433, Validation: 0.8470


In [ ]:
res = model.predict(x_train)

print(res[:10])

94/94 [==============================] - 0s 1ms/step
[[7.1689618e-01]
 [4.1717902e-01]
 [8.3091235e-01]
 [7.9341684e-05]
 [7.4682467e-02]
 [7.3391306e-01]
 [8.4415138e-01]
 [3.5346204e-01]
 [1.4176142e-01]
 [1.6003476e-02]]


จากภาพด้านบน Model จะส่งผลการทำนายเป็นค่าความน่าจะเป็นว่ามีโอกาสที่จะเป็น Class 1 กี่เปอร์เซ็นต์ ซึ่งโดย Default จะทำนายว่าเป็น Class 1 เมื่อค่าความน่าจะเป็นมากกว่า 0.5

In [ ]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(60, input_dim=2, activation='relu', kernel_initializer='he_uniform'))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(1, activation='softmax')) #sigmoid

In [ ]:
opt = tf.keras.optimizers.SGD(lr=0.01, momentum=0.9)
model.compile(loss='binary_crossentropy', optimizer=opt, metrics=['accuracy'])

In [ ]:
his = model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=2000, verbose=1, batch_size = 256)

Epoch 1/2000
12/12 [==============================] - 1s 16ms/step - loss: 0.7512 - accuracy: 0.5027 - val_loss: 0.6878 - val_accuracy: 0.4960
Epoch 2/2000
12/12 [==============================] - 0s 4ms/step - loss: 0.7023 - accuracy: 0.5027 - val_loss: 0.6881 - val_accuracy: 0.4960
Epoch 3/2000
12/12 [==============================] - 0s 6ms/step - loss: 0.7034 - accuracy: 0.5027 - val_loss: 0.6826 - val_accuracy: 0.4960
Epoch 4/2000
12/12 [==============================] - 0s 5ms/step - loss: 0.6987 - accuracy: 0.5027 - val_loss: 0.6772 - val_accuracy: 0.4960
Epoch 5/2000
12/12 [==============================] - 0s 4ms/step - loss: 0.6889 - accuracy: 0.5027 - val_loss: 0.6739 - val_accuracy: 0.4960
Epoch 6/2000
12/12 [==============================] - 0s 5ms/step - loss: 0.6922 - accuracy: 0.5027 - val_loss: 0.6699 - val_accuracy: 0.4960
Epoch 7/2000
12/12 [==============================] - 0s 5ms/step - loss: 0.6823 - accuracy: 0.5027 - val_loss: 0.6667 - val_accuracy: 0.4960
Epoch

In [ ]:
h1 = go.Scatter(y=his.history['loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="loss"
                   )
h2 = go.Scatter(y=his.history['val_loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_loss"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Loss',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

In [ ]:
h1 = go.Scatter(y=his.history['accuracy'],
                    mode="lines", line=dict(
                    width=2,
                    color='blue'),
                    name="acc"
                   )
h2 = go.Scatter(y=his.history['val_accuracy'],
                    mode="lines", line=dict(
                    width=2,
                    color='red'),
                    name="val_acc"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Accuracy',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

In [ ]:
_, train_acc = model.evaluate(x_train, y_train, verbose=0)
_, val_acc = model.evaluate(x_val, y_val, verbose=0)
print('Train: %.4f, Validation: %.4f' % (train_acc, val_acc))

Train: 0.5027, Validation: 0.4960


In [ ]:
res = model.predict(x_train)

print(res[:10])

94/94 [==============================] - 0s 3ms/step
[[0.99999994]
 [0.99999994]
 [0.99999994]
 [0.99999994]
 [0.99999994]
 [0.99999994]
 [0.99999994]
 [0.99999994]
 [0.99999994]
 [0.99999994]]


#Multi-Class Classification Loss Functions

Multi-Class Classification เป็น Model ที่มีการกำหนด Label หรือ Class มากกว่า 2 Class ซึ่งโดยมากจะเริ่มต้นที่ Class 0 จนถึง num_class - 1 ผลลัพธ์ของ Model จะบอกถึงค่าความเชื่อมั่นในการทำนายของแต่ละ Class โดยค่าความเชื่อมั่นทุก Class รวมกันจะเท่ากับ 1.0

เราจะทดลอง Train Multi-Class Classification Model โดยใช้ Categorical Crossentropy Loss ของ Keras Framework ด้วย Dataset ที่ Make ขึ้นจากฟังก์ชัน make_blobs ของ sklearn Library

*เพื่อจะทำให้ Model ทำนายผลออกมาเป็นค่าความเชื่อมั่นที่ทุก Class รวมกันเท่ากับ 1.0 เราจะต้องคอนฟิก Activate Function ใน Output Layer แบบ Softmax ครับ

In [ ]:
def softmax(vector):
    e = np.exp(vector)
    return e/e.sum()

data = [1, 3, 2.5]
result = softmax(data)

print(result)
print(sum(result))

[0.07769558 0.57409699 0.34820743]
1.0


###Categorical Crossentropy Loss

In [ ]:
events = ['Class 0', 'Class 1', 'Class 2']

actual = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
predicted = [[0.9, 0.05, 0.05], [0.05, 0.89, 0.06], [0.05, 0.01, 0.94]]

p = actual[0]
q = predicted[0]

In [ ]:
fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=q, textposition='auto')
])

fig.update_layout(barmode='group', title='Record 1')

In [ ]:
p = actual[1]
q = predicted[1]

fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=q, textposition='auto')
])

fig.update_layout(barmode='group', title='Record 2')

In [ ]:
p = actual[2]
q = predicted[2]

fig = go.Figure(data=[
    go.Bar(name='Actual Probability Distribution', x=events, y=p, text=p, textposition='auto'),
    go.Bar(name='Predicted Probability Distribution', x=events, y=q, text=q, textposition='auto')
])

fig.update_layout(barmode='group', title='Record 3')

Categorical Crossentropy คือ ค่าเฉลี่ยของ Cross-Entropy ที่เกิดจากการแจกแจงความน่าจะเป็น 2 แบบ คือ การแจกแจงความน่าจะเป็นที่เราอยากได้ (Actual) กับการแจกแจงความน่าจะเป็นที่ถูกประมาณโดย Model (Predicted) ของ Class ต่างๆ (Class 0, 1, 2) ดังภาพด้านบน ซึ่งการได้ค่าเฉลี่ยน้อยนั้นดีกว่าการได้ค่าเฉลี่ยมาก โดย Categorical Crossentropy Loss ของ Keras จะมีการใช้งาน Logarithm ฐาน e เช่นเดียวกับ Binary Crossentropy Loss ครับ

In [ ]:
def categorical_cross_entropy(actual, predicted):
    sum = 0.0
    for i in range(len(actual)):
        for j in range(len(actual[i])):
            sum += actual[i][j] * log(1e-15 + predicted[i][j])
    mean = 1.0 / len(actual) * sum
    return -mean

np.around(categorical_cross_entropy(actual, predicted),5)

0.09459

*บวกตัวเลขขนาดเล็ก (1e+15) เพื่อป้องกันการคำนวนค่า log ของ 0

In [ ]:
cce = tf.keras.losses.CategoricalCrossentropy()
np.around(cce(actual, predicted).numpy(),5)

0.09459

###Example

เราจะทดลอง Train Model แบบ Multi-Class Classification จากข้อมูลที่ Make ขึ้นมาด้วยฟังก์ชัน make_blobs ตามขั้นตอนต่อไปนี้

สร้าง Dataset แบบ 3 Class โดยใช้ Function make_blobs ของ Sklearn

In [ ]:
x, y = make_blobs(n_samples=1000, centers=3, n_features=2, cluster_std=2, random_state=2)

เข้ารหัสผลเฉลยแบบ One-hot Encoding เพื่อสร้างการแจกแจงความน่าจะเป็นที่เราอยากได้ (Actual Probability Distribution)

In [ ]:
y = to_categorical(y)

y[:10]

array([[0., 0., 1.],
       [0., 0., 1.],
       [0., 0., 1.],
       [1., 0., 0.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 1.]], dtype=float32)

แบ่งข้อมูลสำหรับ Train และ Test โดยการสุ่มในสัดส่วน 50:50

In [ ]:
x_train, x_val, y_train, y_val = train_test_split(x, y, test_size=0.5, shuffle= True)

x_train.shape, x_val.shape, y_train.shape, y_val.shape

((500, 2), (500, 2), (500, 3), (500, 3))

นำ Dataset ส่วนที่ Train มาแปลงเป็น DataFrame โดยเปลี่ยนชนิดข้อมูลใน Column "class" เป็น String เพื่อทำให้สามารถแสดงสีแบบไม่ต่อเนื่องได้ แล้วนำไป Plot

In [ ]:
y = np.argmax(y_train,axis=1)

x_train_pd = pd.DataFrame(x_train, columns=['x', 'y'])
y_train_pd = pd.DataFrame(y, columns=['class'], dtype='str')

df = pd.concat([x_train_pd, y_train_pd], axis=1)

In [ ]:
fig = px.scatter(df, x="x", y="y", color="class")
fig.show()

นิยาม Model แบบ 3 Class โดยกำหนด Activation Function ใน Layer สุดท้ายเป็น softmax

In [ ]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(60, input_dim=2, activation='relu', kernel_initializer='he_uniform'))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(3, activation='softmax'))

Compile Model โดยกำหนด Loss Function เป็น categorical_crossentropy

In [ ]:
opt =  tf.keras.optimizers.SGD(lr=0.01, momentum=0.9)
model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])

Train Model

In [ ]:
his = model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=1000, verbose=1, batch_size = 128)

Epoch 1/1000
4/4 [==============================] - 0s 51ms/step - loss: 2.7527 - accuracy: 0.4000 - val_loss: 1.9235 - val_accuracy: 0.4260
Epoch 2/1000
4/4 [==============================] - 0s 16ms/step - loss: 2.2621 - accuracy: 0.5000 - val_loss: 1.4684 - val_accuracy: 0.6320
Epoch 3/1000
4/4 [==============================] - 0s 17ms/step - loss: 1.8595 - accuracy: 0.5860 - val_loss: 1.1586 - val_accuracy: 0.6580
Epoch 4/1000
4/4 [==============================] - 0s 15ms/step - loss: 1.4970 - accuracy: 0.6000 - val_loss: 1.1887 - val_accuracy: 0.5340
Epoch 5/1000
4/4 [==============================] - 0s 15ms/step - loss: 1.2655 - accuracy: 0.5640 - val_loss: 0.8092 - val_accuracy: 0.5940
Epoch 6/1000
4/4 [==============================] - 0s 16ms/step - loss: 1.1762 - accuracy: 0.5500 - val_loss: 1.0131 - val_accuracy: 0.6640
Epoch 7/1000
4/4 [==============================] - 0s 11ms/step - loss: 1.2825 - accuracy: 0.5920 - val_loss: 0.9863 - val_accuracy: 0.6600
Epoch 8/1000


his = model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=1000, verbose=1, batch_size = 128)


In [ ]:
h1 = go.Scatter(y=his.history['loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="loss"
                   )
h2 = go.Scatter(y=his.history['val_loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_loss"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Loss',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

Plot Accuracy

In [ ]:
h1 = go.Scatter(y=his.history['accuracy'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="acc"
                   )
h2 = go.Scatter(y=his.history['val_accuracy'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_acc"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Accuracy',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

Evaluate Model

In [ ]:
_, train_acc = model.evaluate(x_train, y_train, verbose=0)
_, val_acc = model.evaluate(x_val, y_val, verbose=0)
print('Train: %.4f, Validation: %.4f' % (train_acc, val_acc))

Train: 0.8300, Validation: 0.8400


In [ ]:
res = model.predict(x_train)

print(np.round(res[:10],3))

16/16 [==============================] - 0s 1ms/step
[[0.003 0.742 0.255]
 [0.001 0.809 0.189]
 [0.006 0.747 0.247]
 [0.999 0.    0.001]
 [0.    0.154 0.846]
 [0.013 0.677 0.311]
 [0.005 0.742 0.253]
 [0.    0.995 0.005]
 [0.999 0.    0.001]
 [0.99  0.    0.01 ]]


จากภาพด้านบน แสดงค่าความเชื่อมั่นในการทำนายของแต่ละ Class (Class 0, 1, 2) จาก Input Data 10 Record โดยค่าความเชื่อมั่นทุก Class รวมกันเท่ากับ 1.0

In [ ]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(60, input_dim=2, activation='relu', kernel_initializer='he_uniform'))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(3, activation='sigmoid'))

In [ ]:
opt =  tf.keras.optimizers.SGD(lr=0.01, momentum=0.9)
model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])

In [ ]:
his = model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=1000, verbose=1, batch_size = 128)

Epoch 1/1000
4/4 [==============================] - 1s 56ms/step - loss: 3.7105 - accuracy: 0.4060 - val_loss: 4.4116 - val_accuracy: 0.3680
Epoch 2/1000
4/4 [==============================] - 0s 10ms/step - loss: 4.0369 - accuracy: 0.4000 - val_loss: 2.7513 - val_accuracy: 0.4880
Epoch 3/1000
4/4 [==============================] - 0s 11ms/step - loss: 2.5371 - accuracy: 0.5320 - val_loss: 1.6598 - val_accuracy: 0.5440
Epoch 4/1000
4/4 [==============================] - 0s 13ms/step - loss: 2.0584 - accuracy: 0.5440 - val_loss: 1.8084 - val_accuracy: 0.6440
Epoch 5/1000
4/4 [==============================] - 0s 16ms/step - loss: 2.0858 - accuracy: 0.6060 - val_loss: 1.4030 - val_accuracy: 0.6480
Epoch 6/1000
4/4 [==============================] - 0s 16ms/step - loss: 1.8020 - accuracy: 0.6000 - val_loss: 1.7984 - val_accuracy: 0.6120
Epoch 7/1000
4/4 [==============================] - 0s 17ms/step - loss: 1.7250 - accuracy: 0.5980 - val_loss: 1.2464 - val_accuracy: 0.6500
Epoch 8/1000


In [ ]:
h1 = go.Scatter(y=his.history['loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="loss"
                   )
h2 = go.Scatter(y=his.history['val_loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_loss"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Loss',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

In [ ]:
h1 = go.Scatter(y=his.history['accuracy'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="acc"
                   )
h2 = go.Scatter(y=his.history['val_accuracy'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_acc"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Accuracy',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

In [ ]:
_, train_acc = model.evaluate(x_train, y_train, verbose=0)
_, val_acc = model.evaluate(x_val, y_val, verbose=0)
print('Train: %.4f, Validation: %.4f' % (train_acc, val_acc))

Train: 0.8220, Validation: 0.8380


In [ ]:
res = model.predict(x_train)

print(np.round(res[:10],3))

16/16 [==============================] - 0s 1ms/step
[[0.002 0.744 0.254]
 [0.002 0.894 0.104]
 [0.004 0.764 0.232]
 [0.999 0.    0.001]
 [0.    0.113 0.886]
 [0.009 0.713 0.278]
 [0.003 0.759 0.238]
 [0.    0.996 0.004]
 [0.998 0.    0.002]
 [0.982 0.    0.018]]


##Sparse Categorical Crossentropy Loss

โดย Default แล้ว ในการสร้าง Model แบบ Multi-Class Classification เราจะคอนฟิก Loss Function เป็น Categorical Crossentropy Loss แต่สำหรับปัญหาบางปัญหาที่มีผลเฉลย (Label) จำนวนมากๆ การเข้ารหัสผลเฉลยแบบ One-hot Encoding เพื่อสร้างการแจกแจงความน่าจะเป็นแบบที่เราอยากได้ (Actual Probability Distribution) จะทำให้สิ้นเปลือง Memory

Sparse Categorical Crossentropy สามารถแก้ปัญหานี้ได้โดยไม่ต้องสร้างผลเฉลยแบบ One-hot Encoding แต่ยังคงมีการคำนวนค่า Cross-entropy ได้เหมือนเดิม ซึ่งในการคอนฟิก Model ให้ใช้งาน Sparse Categorical Crossentropy Loss เรายังคงใช้ Activate Function เป็น Softmax ที่ Output Layer เช่นเดิม

###Example

เราจะทดลอง Train Model แบบ Multi-Class Classification จากข้อมูลที่ Make ขึ้นมาด้วยฟังก์ชัน make_blobs ตามขั้นตอนต่อไปนี้

สร้าง Dataset แบบ 20 Class โดยใช้ Function make_blobs ของ Sklearn

In [ ]:
x, y = make_blobs(n_samples=10000, centers=20, n_features=2, cluster_std=0.1, random_state=2)

one hot encoder

In [ ]:
y = to_categorical(y)

y[:10]

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.,
        0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0.,
        0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
        0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
        0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        1., 0., 0., 0.]], dtype=float32)

แบ่งข้อมูลสำหรับ Train และ Test โดยการสุ่มในสัดส่วน 50:50

In [ ]:
x_train, x_val, y_train, y_val = train_test_split(x, y, test_size=0.5, shuffle= True)

x_train.shape, x_val.shape, y_train.shape, y_val.shape

((5000, 2), (5000, 2), (5000, 20), (5000, 20))

In [ ]:
y_train

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

นำ Dataset ส่วนที่ Train มาแปลงเป็น DataFrame โดยเปลี่ยนชนิดข้อมูลใน Column "class" เป็น String เพื่อทำให้สามารถแสดงสีแบบไม่ต่อเนื่องได้ แล้วนำไป Plot

In [ ]:
# x_train_pd = pd.DataFrame(x_train, columns=['x', 'y'])
# y_train_pd = pd.DataFrame(y_train, columns=['class'], dtype='str')

# df = pd.concat([x_train_pd, y_train_pd], axis=1)

ValueError: ignored

In [ ]:
# df

x         y class
0    -0.507467 -6.049266    15
1     5.917439  1.652530    18
2    -5.637819 -2.970164    14
3     6.904881 -0.143321     8
4    -1.373689 -9.407934     0
...        ...       ...   ...
4995  2.592110  0.721797     5
4996 -5.440291 -7.957742    13
4997 -7.015918  3.915422    19
4998  7.149018 -0.120999     8
4999  0.094972 -2.203637    17

[5000 rows x 3 columns]

In [ ]:
 y_train

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

In [ ]:
fig = px.scatter(df, x="x", y="y", color="class")
fig.show()

นิยาม Model แบบ 20 Class โดยกำหนด Activation Function ใน Layer สุดท้ายเป็น softmax

In [ ]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(60, input_dim=2, activation='relu', kernel_initializer='he_uniform'))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(20, activation='softmax'))

Compile Model โดยกำหนด Loss Function เป็น sparse_categorical_crossentropy

In [ ]:
opt = tf.keras.optimizers.SGD(lr=0.01, momentum=0.9)
model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])

In [ ]:
len( y_train[0])

20

Train Model

In [ ]:
his = model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=1000, verbose=1, batch_size = 128)


Epoch 1/1000
40/40 [==============================] - 1s 6ms/step - loss: 3.3935 - accuracy: 0.3204 - val_loss: 1.1953 - val_accuracy: 0.4868
Epoch 2/1000
40/40 [==============================] - 0s 3ms/step - loss: 1.2832 - accuracy: 0.5106 - val_loss: 0.8375 - val_accuracy: 0.5626
Epoch 3/1000
40/40 [==============================] - 0s 3ms/step - loss: 0.9572 - accuracy: 0.6092 - val_loss: 0.9664 - val_accuracy: 0.6808
Epoch 4/1000
40/40 [==============================] - 0s 4ms/step - loss: 0.9305 - accuracy: 0.6510 - val_loss: 0.7891 - val_accuracy: 0.6672
Epoch 5/1000
40/40 [==============================] - 0s 3ms/step - loss: 0.7697 - accuracy: 0.6980 - val_loss: 0.6906 - val_accuracy: 0.7294
Epoch 6/1000
40/40 [==============================] - 0s 3ms/step - loss: 0.7385 - accuracy: 0.7406 - val_loss: 0.6241 - val_accuracy: 0.7624
Epoch 7/1000
40/40 [==============================] - 0s 4ms/step - loss: 0.6434 - accuracy: 0.7666 - val_loss: 0.6082 - val_accuracy: 0.7590
Epoch 

Plot Loss

In [ ]:
h1 = go.Scatter(y=his.history['loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="loss"
                   )
h2 = go.Scatter(y=his.history['val_loss'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_loss"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Loss',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

Plot Accuracy

In [ ]:
h1 = go.Scatter(y=his.history['accuracy'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='blue'),
                        name="acc"
                   )
h2 = go.Scatter(y=his.history['val_accuracy'],
                    mode="lines",
                    line=dict(
                        width=2,
                        color='red'),
                        name="val_acc"
                   )

data = [h1,h2]
layout1 = go.Layout(title='Accuracy',
                   xaxis=dict(title='epochs'),
                   yaxis=dict(title=''))
fig1 = go.Figure(data = data, layout=layout1)
plotly.offline.iplot(fig1)

Evaluate Model

In [ ]:
_, train_acc = model.evaluate(x_train, y_train, verbose=0)
_, val_acc = model.evaluate(x_val, y_val, verbose=0)
print('Train: %.4f, Validation: %.4f' % (train_acc, val_acc))

Train: 0.9996, Validation: 0.9998


In [ ]:
res = model.predict(x_train)

print(np.round(res[:1],3))

157/157 [==============================] - 0s 997us/step
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]]


จากภาพด้านบน แสดงค่าความเชื่อมั่นในการทำนายของแต่ละ Class (Class 0 - 19) จาก Input Data 1 Record โดยค่าความเชื่อมั่นทุก Class รวมกันเท่ากับ 1.0